doc: nrf54h: Add SDFW update description (#18383)
nrfconnect/sdk-nrf: doc/nrf/app_dev/device_guides/working_with_nrf/nrf54h/ug_nrf54h20_suit_sdfw_sdrfw_update.rst

.. _ug_nrf54h20_suit_sdfw_sdrfw_update:

Secure Domain Firmware Update
#############################

.. contents::
   :local:
   :depth: 2

Both the Secure Domain Firmware (SDFW) and the recovery image for the Secure Domain Firmware are launched on a core responsible for establishing the Root of Trust as a Secure Domain.
Due to security requirements, the update process for these binaries differs significantly from regular updates in local domains.

To ensure that trusted execution can be established on the device, a valid recovery image for the Secure Domain Firmware must be present on the device before performing a Secure Domain Firmware update.

> An update is initiated only if the update candidate is verified by the Secure Domain ROM using a key that matches the requested update type.

Reviewer: From end-user perspective: it would be great to mention that the bundle that the user programs already contains both SDFW and SDFW Recovery images, so there it no need to look for the magic "Recovery" binary.

Author: Added a note.

Implementation
**************

SUIT manifests use components to define executable or updatable entities.
On a given platform, components may have different characteristics that require varied approaches for launching or updating (for example, execution in place instead of copying to RAM).

As described in :ref:`suit_component_types`, a special SUIT component type (``SOC_SPEC``) describes components that are controlled by Nordic Semiconductor.

For the nRF54H20 SoC, there are two components of this type: one for the SDFW and the other for the SDFW recovery image.
These are identified by component IDs 1 and 2, respectively:

* ``SOC_SPEC/1`` - Secure Domain Firmware
* ``SOC_SPEC/2`` - Secure Domain Firmware recovery image

As shown in :ref:`suit_default_manifest_topology_for_the_nrf54h20_soc`, both components are included in a single SUIT manifest.
This manifest is managed by the ``Nordic top`` manifest, which is, in turn, controlled by the ``Root`` manifest.

Update process
**************

The update process begins by attempting to update the Secure Domain Firmware (SDFW) recovery image.
If this update fails, the process still proceeds with updating the SDFW.

> To prevent ambiguity if one slot update fails while the other succeeds, the manifest version reflects the version of the Secure Domain Firmware.

Reviewer: One important note for the end user is that every update attempt results in the system reboot, and as a result, multiple passes through the root manifest during update sequences are expected.

Author: Added a note.

Reviewer 2: Would be good to have one common naming convention for Recovery. As on the diagrams we call it SDRFW I would suggest to call it "Secure Domain Recovery Firmware" instead of "Secure Domain Firmware Recovery"

.. figure:: images/nrf54h20_suit_sdfw_sdrfw_update_flow.png
   :alt: Secure Domain Firmware and Secure Domain Firmware Recovery update flow

>    Update flow for the Secure Domain Firmware and the Secure Domain Firmware recovery image

Reviewer 2: Maybe we can mention in the end that some more information about strategy, configuration and update of Nordic components can be found under https://docs.nordicsemi.com/bundle/ncs-latest/page/nrf/app_dev/device_guides/working_with_nrf/nrf54h/ug_nrf54h20_suit_soc_binaries.html

Author: Done

> .. _ug_nrf54h20_suit_sdfw_sdrfw_update:

Reviewer 3: The image also needs to be remade

Author: What would you like to change in the image?

Reviewer 3: It will be updated in a separate PR

> #############################

Reviewer 3: ```suggestion
```

Secure Domain Firmware Update
#############################

.. contents::
   :local:
   :depth: 2

Background
**********

> Both Secure Domain Firmware and Secure Domain Firmware Recovery are launched on a core responsible for establishing Root of Trust as a Secure Domain. Due to security requirements update process of these binaries differs significantly from regular updates of Local Domains.
> In order to ensure that a trusted execution can be established on the device, a valid Secure Domain Firmware Recovery has to be present on the device prior to performing an Secure Domain Firmware update.
> An update is started only if the update candidate can be verified by Secure Domain ROM using a key matching the requested update type.

Reviewer 3: ```suggestion
Both the Secure Domain Firmware (SDFW) and the recovery image for the Secure Domain Firmware are launched on a core responsible for establishing the Root of Trust as a Secure Domain.
Due to security requirements, the update process for these binaries differs significantly from regular updates in Local Domains.

To ensure that trusted execution can be established on the device, a valid recovery image for the Secure Domain Firmware must be present on the device before performing a Secure Domain Firmware update.
An update is initiated only if the update candidate is verified by the Secure Domain ROM using a key that matches the requested update type.

```

> SUIT attempts to provide flexibility with regard to launching and updating firmware while allowing certain parts of the code base to remain unmodifiable for end users. Although there are various implementations possible, a platform-specific layer that maps logical firmware blocks to their corresponding memory ranges of a device is a straightforward approach for simple systems. The manifest(s) can describe actions needed to launch or update such firmware block.
> For certain areas, nrf54 family uses such approach, however Secure Domain Firmware and Secure Domain Firmware Recovery cannot be simply copied to destination address in order to replace the old firmware - its design relies on Secure Domain ROM launching and updating the binaries.

Reviewer 3: ```suggestion
The SUIT framework allows flexible firmware launching and updating while keeping parts of the codebase unmodifiable by end users.
A common approach is to use a platform-specific layer that maps firmware blocks to device memory ranges, simplifying updates for basic systems.
Manifests can describe actions required to launch or update each firmware block.
In the nRF54H20, this approach is used in some areas, but the SDFW and the recovery image of the SDFW require the Secure Domain ROM to handle launching and updates, rather than direct copying to a destination address.
```

Implementation
**************

> SUIT manifests utilize a concept of components that describe executable/updatable entities. For a given platform there can be a variety of components with different characteristics that require different approach for launching or updating given component (e.g. execution in place versus copying to RAM, etc.).

Reviewer 3: ```suggestion
SUIT manifests use components to define executable or updatable entities.
On a given platform, components may have different characteristics that require varied approaches for launching or updating (for example, execution in place instead of copying to RAM).

```

> As described in :ref:`suit_component_types`, there is a special SUIT component type ``SOC_SPEC``
> available for describing components that are controlled by Nordic.

Reviewer 3: ```suggestion
As described in :ref:`suit_component_types`, a special SUIT component type (``SOC_SPEC``) describes components that are controlled by Nordic Semiconductor.
```

> For nRF54H20, there are two components of this type, one for Secure Domain Firmware and the

Reviewer 3: ```suggestion
For the nRF54H20 SoC, there are two components of this type: one for the SDFW and the other for the SDFW recovery image.
```

> other one for Secure Domain Recovery Firmware.

Reviewer 3: ```suggestion
```

> They are distinguished by component IDs of 1 and 2, respectively:
> 
> * ``SOC_SPEC/1`` - Secure Domain Firmware
> 
> * ``SOC_SPEC/2`` - Secure Domain Firmware Recovery

Reviewer 3: ```suggestion
These are identified by component IDs 1 and 2, respectively:

* ``SOC_SPEC/1`` - Secure Domain Firmware
* ``SOC_SPEC/2`` - Secure Domain Firmware recovery image
```

> As shown in :ref:`suit_default_manifest_topology_for_the_nrf54h20_soc` both components are part of a
> single SUIT manifest. The manifest is controlled by ``Nordic top`` manifest, which in turn is
> controlled by the ``Root`` manifest.

Reviewer 3: ```suggestion
As shown in :ref:`suit_default_manifest_topology_for_the_nrf54h20_soc`, both components are included in a single SUIT manifest.
This manifest is managed by the ``Nordic top`` manifest, which is, in turn, controlled by the ``Root`` manifest.
```

Update process
**************

> Design decision that Secure Domain Firmware and Secure Domain Firmware Recovery are controlled by a single manifest has impact on how these two slots are updated.
> When update of one of the slots fails while the other one succeeds, it could lead to ambiguity of system state.
> Because of that, it was decided that manifest version will reflect Secure Domain Firmware version, as the more crucial slot of the two.
> Secure Domain Firmware Recovery update is attempted first and when it fails, it should not prevent attempt of Secure Domain Firmware update.

Reviewer 3: ```suggestion
The update process begins by attempting to update the Secure Domain Firmware (SDFW) recovery image.
If this update fails, the process still proceeds with updating the SDFW.
To prevent ambiguity if one slot update fails while the other succeeds, the manifest version reflects the version of the Secure Domain Firmware.
```

.. figure:: images/nrf54h20_suit_sdfw_sdrfw_update_flow.png
   :alt: Secure Domain Firmware and Secure Domain Firmware Recovery update flow

>    Secure Domain Firmware and Secure Domain Firmware Recovery update flow

Reviewer 3: ```suggestion
   Update flow for the Secure Domain Firmware and the Secure Domain Firmware recovery image 
```

.. _ug_nrf54h20_suit_sdfw_sdrfw_update:

Secure Domain Firmware Update
#############################

.. contents::
   :local:
   :depth: 2

> Background
> **********

Reviewer 3: ```suggestion
```

The first section does not need a title.

Both the Secure Domain Firmware (SDFW) and the recovery image for the Secure Domain Firmware are launched on a core responsible for establishing the Root of Trust as a Secure Domain.
Due to security requirements, the update process for these binaries differs significantly from regular updates in Local Domains.

To ensure that trusted execution can be established on the device, a valid recovery image for the Secure Domain Firmware must be present on the device before performing a Secure Domain Firmware update.
An update is initiated only if the update candidate is verified by the Secure Domain ROM using a key that matches the requested update type.

> 
> 
> 
> The SUIT framework allows flexible firmware launching and updating while keeping parts of the codebase unmodifiable by end users.

Reviewer 3: ```suggestion

The SUIT framework allows flexible firmware launching and updating while keeping parts of the codebase unmodifiable by end users.
```

A common approach is to use a platform-specific layer that maps firmware blocks to device memory ranges, simplifying updates for basic systems.
Manifests can describe actions required to launch or update each firmware block.
In the nRF54H20, this approach is used in some areas, but the SDFW and the recovery image of the SDFW require the Secure Domain ROM to handle launching and updates, rather than direct copying to a destination address.

Implementation
**************

SUIT manifests use components to define executable or updatable entities.
On a given platform, components may have different characteristics that require varied approaches for launching or updating (for example, execution in place instead of copying to RAM).

> 
> 
> As described in :ref:`suit_component_types`, a special SUIT component type (``SOC_SPEC``) describes components that are controlled by Nordic Semiconductor.

Reviewer 3: ```suggestion

As described in :ref:`suit_component_types`, a special SUIT component type (``SOC_SPEC``) describes components that are controlled by Nordic Semiconductor.
```
